File: core/src/com/megaman/game/entities/impl/enemies/impl/MagFly.java
package com.megaman.game.entities.impl.enemies.impl;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.ObjectMap;
import com.megaman.game.ConstKeys;
import com.megaman.game.MegamanGame;
import com.megaman.game.ViewVals;
import com.megaman.game.animations.Animation;
import com.megaman.game.animations.AnimationComponent;
import com.megaman.game.assets.TextureAsset;
import com.megaman.game.entities.utils.damage.DamageNegotiation;
import com.megaman.game.entities.utils.damage.Damager;
import com.megaman.game.entities.utils.faceable.Faceable;
import com.megaman.game.entities.utils.faceable.Facing;
import com.megaman.game.entities.impl.enemies.Enemy;
import com.megaman.game.entities.impl.explosions.impl.ChargedShotExplosion;
import com.megaman.game.entities.impl.megaman.Megaman;
import com.megaman.game.entities.impl.projectiles.Projectile;
import com.megaman.game.entities.impl.projectiles.impl.Bullet;
import com.megaman.game.entities.impl.projectiles.impl.ChargedShot;
import com.megaman.game.entities.impl.projectiles.impl.Fireball;
import com.megaman.game.health.HealthVals;
import com.megaman.game.shapes.ShapeComponent;
import com.megaman.game.shapes.ShapeHandle;
import com.megaman.game.shapes.ShapeUtils;
import com.megaman.game.sprites.SpriteComponent;
import com.megaman.game.sprites.SpriteHandle;
import com.megaman.game.updatables.UpdatableComponent;
import com.megaman.game.utils.enums.Position;
import com.megaman.game.utils.interfaces.UpdateFunc;
import com.megaman.game.utils.objs.Timer;
import com.megaman.game.world.*;
import lombok.Getter;
import lombok.Setter;

import java.util.HashMap;
import java.util.Map;

import static com.badlogic.gdx.graphics.Color.GRAY;

public class MagFly extends Enemy implements Faceable {

    private static final float FORCE_FLASH_DURATION = .1f;
    private static final float X_VEL_NORMAL = 3f;
    private static final float X_VEL_SLOW = 1f;
    private static final float PULL_FORCE_X = 6f;
    private static final float PULL_FORCE_Y = 68f;

    private static TextureRegion magFlyReg;

    private final Sprite sprite;
    private final Timer forceFlashTimer;

    private Fixture forceFixture;

    private boolean flash;
    @Getter
    @Setter
    private Facing facing;

    public MagFly(MegamanGame game) {
        super(game, BodyType.DYNAMIC);
        if (magFlyReg == null) {
            magFlyReg = game.getAssMan().getTextureRegion(TextureAsset.ENEMIES_1, "MagFly");
        }
        defineBody();
        sprite = new Sprite();
        forceFlashTimer = new Timer(FORCE_FLASH_DURATION);
        if (MegamanGame.DEBUG) {
            putComponent(shapeComponent());
        }
        putComponent(spriteComponent());
        putComponent(animationComponent());
    }

    @Override
    public void init(Rectangle bounds, ObjectMap<String, Object> data) {
        Vector2 spawn = ShapeUtils.getCenterPoint(bounds);
        body.bounds.setCenter(spawn);
    }

    @Override
    protected Map<Class<? extends Damager>, DamageNegotiation> defineDmgNegs() {
        return new HashMap<>() {{
            put(Bullet.class, new DamageNegotiation(10));
            put(Fireball.class, new DamageNegotiation(HealthVals.MAX_HEALTH));
            put(ChargedShot.class, new DamageNegotiation(HealthVals.MAX_HEALTH));
            put(ChargedShotExplosion.class, new DamageNegotiation(HealthVals.MAX_HEALTH));
        }};
    }

    protected void defineBody() {
        body.bounds.setSize(WorldVals.PPM);
        Fixture bodyFixture = new Fixture(this, FixtureType.BODY, new Rectangle().setSize(WorldVals.PPM));
        body.add(bodyFixture);
        Fixture leftFixture = new Fixture(this, FixtureType.SIDE, new Rectangle().setSize(.1f * WorldVals.PPM));
        leftFixture.putUserData(ConstKeys.SIDE, ConstKeys.LEFT);
        leftFixture.offset.x = -.6f * WorldVals.PPM;
        leftFixture.offset.y = -.4f * WorldVals.PPM;
        body.add(leftFixture);
        Fixture rightFixture = new Fixture(this, FixtureType.SIDE, new Rectangle().setSize(.1f * WorldVals.PPM));
        rightFixture.putUserData(ConstKeys.SIDE, ConstKeys.RIGHT);
        rightFixture.offset.x = .6f * WorldVals.PPM;
        rightFixture.offset.y = -.4f * WorldVals.PPM;
        body.add(rightFixture);
        Fixture forceFixture = new Fixture(this, FixtureType.FORCE,
                new Rectangle().setSize(WorldVals.PPM / 2f, ViewVals.VIEW_HEIGHT * WorldVals.PPM));
        forceFixture.offset.y = -ViewVals.VIEW_HEIGHT * WorldVals.PPM / 2f;

        // TODO: test

        /*
        Function<Fixture, Vector2> forceFunc = f -> {
            if (f.entity instanceof Enemy || (f.entity instanceof Megaman m && m.isDamaged())) {
                return Vector2.Zero;
            }
            if (f.entity instanceof Projectile p) {
                p.owner = null;
            }
            float x = PULL_FORCE_X * WorldVals.PPM;
            if (is(Facing.LEFT)) {
                x *= -1f;
            }
            float y = PULL_FORCE_Y * WorldVals.PPM;
            return new Vector2(x, y);
        };
         */

        /*
        Function<Float, Vector2> forceFunc = delta -> {
            float x = PULL_FORCE_X * WorldVals.PPM;
            if (is(Facing.LEFT)) {
                x *= -1f;
            }
            float y = PULL_FORCE_Y * WorldVals.PPM;
            return new Vector2(x, y).scl(delta);
        };
         */

        UpdateFunc<Fixture, Vector2> forceFunc = (f, delta) -> {
            if (f.entity instanceof Enemy || (f.entity instanceof Megaman m && m.isDamaged())) {
                return Vector2.Zero;
            }
            if (f.entity instanceof Projectile p) {
                p.owner = null;
            }
            float x = PULL_FORCE_X * WorldVals.PPM;
            if (is(Facing.LEFT)) {
                x *= -1f;
            }
            float y = PULL_FORCE_Y * WorldVals.PPM;
            return new Vector2(x, y).scl(delta);
        };

        forceFixture.putUserData(ConstKeys.FUNCTION, forceFunc);

        body.add(forceFixture);
        this.forceFixture = forceFixture;
        Fixture damageableFixture = new Fixture(this, FixtureType.DAMAGEABLE, new Rectangle().setSize(WorldVals.PPM));
        body.add(damageableFixture);
        Fixture damagerFixture = new Fixture(this, FixtureType.DAMAGER, new Rectangle().setSize(.85f * WorldVals.PPM));
        body.add(damagerFixture);
    }

    @Override
    protected void defineUpdateComponent(UpdatableComponent c) {
        super.defineUpdateComponent(c);
        c.add(delta -> {
            forceFlashTimer.update(delta);
            if (forceFlashTimer.isFinished()) {
                flash = !flash;
                forceFlashTimer.reset();
            }
            boolean slow = isMegamanOverlappingForceFixture();
            if (!slow && !isMegamanAbove() && !doFacingAndMMDirMatch()) {
                setFacing(isMegamanRight() ? Facing.RIGHT : Facing.LEFT);
            }
            if ((is(Facing.LEFT) && is(BodySense.SIDE_TOUCHING_BLOCK_LEFT)) ||
                    (is(Facing.RIGHT) && is(BodySense.SIDE_TOUCHING_BLOCK_RIGHT))) {
                body.velocity.x = 0f;
            } else {
                body.velocity.x = (slow ? X_VEL_SLOW : X_VEL_NORMAL) * WorldVals.PPM;
                if (is(Facing.LEFT)) {
                    body.velocity.x *= -1f;
                }
            }
        });
    }

    private boolean isMegamanRight() {
        return game.getMegaman().body.isRightOf(body);
    }

    private boolean isMegamanAbove() {
        return game.getMegaman().body.isAbove(body);
    }

    private boolean doFacingAndMMDirMatch() {
        return (isMegamanRight() && is(Facing.RIGHT)) || (!isMegamanRight() && is(Facing.LEFT));
    }

    private boolean isMegamanOverlappingForceFixture() {
        Body megamanBody = game.getMegaman().body;
        return ShapeUtils.overlaps(forceFixture.shape, megamanBody.bounds);
    }

    private SpriteComponent spriteComponent() {
        sprite.setSize(1.5f * WorldVals.PPM, 1.5f * WorldVals.PPM);
        SpriteHandle h = new SpriteHandle(sprite, 4);
        h.updatable = delta -> {
            h.setPosition(body.bounds, Position.CENTER);
            sprite.setFlip(is(Facing.LEFT), false);
        };
        return new SpriteComponent(h);
    }

    private AnimationComponent animationComponent() {
        return new AnimationComponent(sprite, new Animation(magFlyReg, 2, .1f));
    }

    private ShapeComponent shapeComponent() {
        ShapeHandle shapeHandle = new ShapeHandle();
        shapeHandle.setShapeSupplier(() -> forceFixture.shape);
        shapeHandle.setDoRenderSupplier(() -> flash);
        shapeHandle.setColorSupplier(() -> GRAY);
        return new ShapeComponent(shapeHandle);
    }

}

File: core/src/main/kotlin/entities/enemies/MagFly.kt
package com.megaman.maverick.game.entities.enemies

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.utils.Array
import com.mega.game.engine.animations.Animation
import com.mega.game.engine.animations.AnimationsComponent
import com.mega.game.engine.animations.Animator
import com.mega.game.engine.common.enums.Facing
import com.mega.game.engine.common.extensions.getTextureRegion
import com.mega.game.engine.common.extensions.objectMapOf
import com.mega.game.engine.common.interfaces.IFaceable
import com.mega.game.engine.common.objects.Properties
import com.mega.game.engine.common.objects.pairTo
import com.mega.game.engine.common.shapes.GameRectangle
import com.mega.game.engine.common.time.Timer
import com.mega.game.engine.damage.IDamager
import com.mega.game.engine.drawables.shapes.DrawableShapesComponent
import com.mega.game.engine.drawables.shapes.IDrawableShape
import com.mega.game.engine.drawables.sprites.GameSprite
import com.mega.game.engine.drawables.sprites.SpritesComponent
import com.mega.game.engine.drawables.sprites.setCenter
import com.mega.game.engine.drawables.sprites.setSize
import com.mega.game.engine.updatables.UpdatablesComponent
import com.mega.game.engine.world.body.Body
import com.mega.game.engine.world.body.BodyComponent
import com.mega.game.engine.world.body.BodyType
import com.mega.game.engine.world.body.Fixture
import com.megaman.maverick.game.ConstKeys
import com.megaman.maverick.game.ConstVals
import com.megaman.maverick.game.MegamanMaverickGame
import com.megaman.maverick.game.assets.TextureAsset
import com.megaman.maverick.game.damage.DamageNegotiation
import com.megaman.maverick.game.damage.dmgNeg
import com.megaman.maverick.game.entities.contracts.AbstractEnemy
import com.megaman.maverick.game.entities.contracts.AbstractProjectile
import com.megaman.maverick.game.entities.explosions.ChargedShotExplosion
import com.megaman.maverick.game.entities.megaman.Megaman
import com.megaman.maverick.game.entities.projectiles.Bullet
import com.megaman.maverick.game.entities.projectiles.ChargedShot
import com.megaman.maverick.game.entities.projectiles.Fireball
import com.megaman.maverick.game.utils.VelocityAlteration

import com.megaman.maverick.game.world.body.*
import kotlin.reflect.KClass

class MagFly(game: MegamanMaverickGame) : AbstractEnemy(game), IFaceable {

    companion object {
        const val TAG = "MagFly"
        private const val FORCE_FLASH_DURATION = 0.1f
        private const val X_VEL_NORMAL = 3f
        private const val X_VEL_SLOW = 1f
        private const val PULL_FORCE_X = 6f
        private const val PULL_FORCE_Y = 68f

        private var magFlyReg: TextureRegion? = null
    }

    override val damageNegotiations = objectMapOf<KClass<out IDamager>, DamageNegotiation>(
        Bullet::class pairTo dmgNeg(15),
        Fireball::class pairTo dmgNeg(ConstVals.MAX_HEALTH),
        ChargedShot::class pairTo dmgNeg(ConstVals.MAX_HEALTH),
        ChargedShotExplosion::class pairTo dmgNeg(ConstVals.MAX_HEALTH)
    )
    override var facing = Facing.RIGHT

    private val forceFlashTimer = Timer(FORCE_FLASH_DURATION)
    private var flash = false
    private lateinit var forceFixture: Fixture

    override fun init() {
        super.init()
        if (magFlyReg == null) magFlyReg = game.assMan.getTextureRegion(TextureAsset.ENEMIES_1.source, "MagFly")
        addComponent(defineAnimationsComponent())
    }

    override fun onSpawn(spawnProps: Properties) {
        super.onSpawn(spawnProps)
        val spawn = spawnProps.get(ConstKeys.BOUNDS, GameRectangle::class)!!.getCenter()
        body.setCenter(spawn)
    }

    override fun defineBodyComponent(): BodyComponent {
        val body = Body(BodyType.DYNAMIC)
        body.setSize(ConstVals.PPM.toFloat())

        val debugShapes = Array<() -> IDrawableShape?>()
        debugShapes.add { body.getBounds() }

        val bodyFixture = Fixture(body, FixtureType.BODY, GameRectangle().setSize(ConstVals.PPM.toFloat()))
        body.addFixture(bodyFixture)
        debugShapes.add { bodyFixture.getShape() }

        val leftFixture = Fixture(body, FixtureType.SIDE, GameRectangle().setSize(0.1f * ConstVals.PPM))
        leftFixture.putProperty(ConstKeys.SIDE, ConstKeys.LEFT)
        leftFixture.offsetFromBodyAttachment.x = -0.6f * ConstVals.PPM
        leftFixture.offsetFromBodyAttachment.y = 0.4f * ConstVals.PPM
        body.addFixture(leftFixture)
        debugShapes.add { leftFixture.getShape() }

        val rightFixture = Fixture(body, FixtureType.SIDE, GameRectangle().setSize(0.1f * ConstVals.PPM))
        rightFixture.putProperty(ConstKeys.SIDE, ConstKeys.RIGHT)
        rightFixture.offsetFromBodyAttachment.x = 0.6f * ConstVals.PPM
        rightFixture.offsetFromBodyAttachment.y = 0.4f * ConstVals.PPM
        body.addFixture(rightFixture)
        debugShapes.add { rightFixture.getShape() }

        forceFixture = Fixture(
            body,
            FixtureType.FORCE,
            GameRectangle().setSize(ConstVals.PPM / 2f, ConstVals.VIEW_HEIGHT * ConstVals.PPM)
        )
        forceFixture.offsetFromBodyAttachment.y = -ConstVals.VIEW_HEIGHT * ConstVals.PPM / 2f
        forceFixture.setVelocityAlteration { fixture, delta ->
            val entity = fixture.getEntity()
            if (entity is AbstractEnemy || (entity is Megaman && entity.damaged)) return@setVelocityAlteration VelocityAlteration.addNone()
            if (entity is AbstractProjectile) entity.owner = null

            val x = PULL_FORCE_X * ConstVals.PPM * facing.value
            val y = PULL_FORCE_Y * ConstVals.PPM * delta

            return@setVelocityAlteration VelocityAlteration.add(x, y)
        }
        body.addFixture(forceFixture)
        forceFixture.getShape().color = Color.YELLOW
        debugShapes.add { forceFixture.getShape() }

        val damageableFixture = Fixture(body, FixtureType.DAMAGEABLE, GameRectangle().setSize(ConstVals.PPM.toFloat()))
        body.addFixture(damageableFixture)
        damageableFixture.getShape().color = Color.PURPLE
        debugShapes.add { damageableFixture.getShape() }

        val damagerFixture = Fixture(body, FixtureType.DAMAGER, GameRectangle().setSize(0.95f * ConstVals.PPM))
        body.addFixture(damagerFixture)
        damagerFixture.getShape().color = Color.RED
        debugShapes.add { damagerFixture.getShape() }

        addComponent(DrawableShapesComponent(debugShapeSuppliers = debugShapes, debug = true))

        return BodyComponentCreator.create(this, body)
    }

    override fun defineSpritesComponent(): SpritesComponent {
        val sprite = GameSprite()
        sprite.setSize(1.5f * ConstVals.PPM)
        val spritesComponent = SpritesComponent(sprite)
        spritesComponent.putUpdateFunction { _, _sprite ->
            _sprite.hidden = damageBlink
            _sprite.setCenter(body.getCenter())
            _sprite.setFlip(facing == Facing.LEFT, false)
        }
        return spritesComponent
    }

    override fun defineUpdatablesComponent(updatablesComponent: UpdatablesComponent) {
        super.defineUpdatablesComponent(updatablesComponent)
        updatablesComponent.add {
            forceFlashTimer.update(it)
            if (forceFlashTimer.isFinished()) {
                flash = !flash
                forceFlashTimer.reset()
            }

            val slow = megaman().body.overlaps(forceFixture.getShape() as Rectangle)

            if (!slow && megaman().body.getY() < body.getY() && !facingAndMMDirMatch()) facing =
                if (megaman().body.getX() > body.getX()) Facing.RIGHT else Facing.LEFT

            if ((facing == Facing.LEFT && body.isSensing(BodySense.SIDE_TOUCHING_BLOCK_LEFT)) || (facing == Facing.RIGHT && body.isSensing(
                    BodySense.SIDE_TOUCHING_BLOCK_RIGHT
                ))
            ) body.physics.velocity.x = 0f
            else body.physics.velocity.x = (if (slow) X_VEL_SLOW else X_VEL_NORMAL) * ConstVals.PPM * facing.value
        }
    }

    private fun defineAnimationsComponent(): AnimationsComponent {
        val animation = Animation(magFlyReg!!, 1, 2, 0.1f, true)
        val animator = Animator(animation)
        return AnimationsComponent(this, animator)
    }

    private fun facingAndMMDirMatch() =
        (game.megaman.body.getX() > body.getX() && facing == Facing.RIGHT) || (game.megaman.body.getX() < body.getX() && facing == Facing.LEFT)
}
